modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/DynamicOptionsParameterDeclarationEnricher.java

```diff
@@ -0,0 +1,231 @@
+    Optional<ImplementingTypeModelProperty> implementingType =
+        extractImplementingTypeProperty(extensionLoadingContext.getExtensionDeclarer().getDeclaration());
+
+    typeLoader = ExtensionsTypeLoaderFactory.getDefault().createTypeLoader(currentThread().getContextClassLoader());
+
+    if (implementingType.isPresent()) {
+      new IdempotentDeclarationWalker() {
+
+        @Override
+        public void onSource(SourceDeclaration declaration) {
+          enrichContainerModel(declaration, declaration::addModelProperty);
+        }
+
+        @Override
+        public void onOperation(OperationDeclaration declaration) {
+          enrichContainerModel(declaration, declaration::addModelProperty);
+        }
+
+        @Override
+        protected void onConfiguration(ConfigurationDeclaration declaration) {
+          enrichContainerModel(declaration, declaration::addModelProperty);
+        }
+
+        @Override
+        protected void onConnectionProvider(ConnectionProviderDeclaration declaration) {
+          enrichContainerModel(declaration, declaration::addModelProperty);
+        }
+      }.walk(extensionLoadingContext.getExtensionDeclarer().getDeclaration());
+    }
+  }
+
+  private void enrichContainerModel(ParameterizedDeclaration<?> declaration, Consumer<ModelProperty> enrichableModel) {
+    List<ParameterDeclaration> allParameters = declaration.getAllParameters();
+
+    Map<ParameterGroupDeclaration, Class<? extends OptionsResolver>> dynamicGroupOptions = introspectParameterGroups(declaration.getParameterGroups());
+    Map<ParameterDeclaration, Class<? extends OptionsResolver>> dynamicOptions = introspectParameters(allParameters);
+
+    OptionsModelPropertyBuilder builder = OptionsModelProperty.builder();
+
+    dynamicOptions.forEach((paramDeclaration, resolverClass) -> enrichParameter(paramDeclaration.getName(), resolverClass, paramDeclaration, builder::forParameter));
+    dynamicGroupOptions.forEach((paramDeclaration, resolverClass) -> enrichParameter(paramDeclaration.getName(), resolverClass, paramDeclaration, builder::forParameterGroup));
+
+    enrichableModel.accept(builder.build());
+  }
+
+  private void enrichParameter(String parameterName, Class<? extends OptionsResolver> resolverClass, BaseDeclaration declaration, OptionsModelPropertyAlgo optionsModelPropertyAlgo) {
+    OptionsResolverFactoryModelPropertyBuilder propertyBuilder =
+        OptionsResolverFactoryModelProperty.builder(parameterName, resolverClass);
+    ParameterizableTypeWrapper resolverClassWrapper = new ParameterizableTypeWrapper(resolverClass);
+
+    List<ExtensionParameter> resolverParameters = resolverClassWrapper.getParametersAnnotatedWith(Parameter.class);
+    resolverParameters.forEach(param -> propertyBuilder
+        .withInjectableParameter(param.getName(), param.getMetadataType(typeLoader), param.isRequired()));
+
+    injectConnection(propertyBuilder, resolverClassWrapper);
```

Comment:
nothing should be injected, everything needs to be communicated through immutable, discardable contexts. Otherwise this wouldn't work with reconnection, connection management, etc